package/upgrade/upgrade_node.sh

```diff
@@ -186,14 +186,42 @@ shutdown_non_migrate_able_vms()
     done
 
   # VMs with nodeAffinity
+  # Skip only when the VMs have other places to go
+  local node_labels=""
+  local node_count=0
   kubectl get vmi -A -l kubevirt.io/nodeName=$HARVESTER_UPGRADE_NODE_NAME -o json |
     jq -r '.items[] | select(.spec.affinity.nodeAffinity != null) | [.metadata.name, .metadata.namespace] | @tsv' |
     while IFS=$'\t' read -r name namespace; do
       if [ -z "$name" ]; then
         break
       fi
-      echo "Stop ${namespace}/${name}"
-      virtctl stop $name -n $namespace
+      node_labels=$(kubectl -n "$namespace" get vmi "$name" -o yaml |
+        yq '.spec.affinity.nodeAffinity.*.nodeSelectorTerms[].matchExpressions[] | select(.operator=="In" and .values[] == "true") | .key')
+      for nl in $node_labels; do
+        # For nodes considered "condidates" for the VM to live-migrate, they should
+        # 1. match the same labels as nodeAffinity described
+        # 2. not be the node that the VM currently runs on
+        # 3. be schedulable at the time
+        node_count=$(kubectl get nodes -o yaml |
+                nl="$nl" n=$HARVESTER_UPGRADE_NODE_NAME yq '.items[] | select(.metadata.labels.[env(nl)] == "true" and .metadata.name != env(n) and .spec.unschedulable == null) | .metadata.name' | wc -l)
```

Comment:
As a side note, `yq` array evaluation doesn't output `null' if there is no match:
```
$ echo '' | yq '.items[]'                                                                                                                                                            

$ echo '' | yq '.items'
null
```